pkg/cloudprovider/aws/amifamily/ami.go

```diff
@@ -19,30 +19,170 @@ import (
+	if providerRef != nil {
+		var ant v1alpha1.AWSNodeTemplate
+		if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: providerRef.Name}, &ant); err != nil {
+			return amiRequirements, fmt.Errorf("retrieving provider reference, %w", err)
+		}
+		if len(ant.Spec.AMISelector) == 0 {
+			return amiRequirements, nil
+		}
+		return p.selectAMIs(ctx, ant.Spec.AMISelector)
+	}
+	return amiRequirements, nil
+}
+
+func (p *AMIProvider) selectAMIs(ctx context.Context, amiSelector map[string]string) (map[string]scheduling.Requirements, error) {
+	ec2AMIs, err := p.fetchAMIsFromEC2(ctx, amiSelector)
+	if err != nil {
+		return nil, err
+	}
+	if len(ec2AMIs) == 0 {
+		return nil, fmt.Errorf("no amis exist given constraints")
+	}
+	var amiIDs = map[string]scheduling.Requirements{}
+	for _, ec2AMI := range ec2AMIs {
+		amiIDs[*ec2AMI.ImageId] = p.getRequirementsFromImage(ec2AMI)
+	}
+	return amiIDs, nil
+}
+
+func (p *AMIProvider) fetchAMIsFromEC2(ctx context.Context, amiSelector map[string]string) ([]*ec2.Image, error) {
+	filters := getFilters(amiSelector)
+	hash, err := hashstructure.Hash(filters, hashstructure.FormatV2, nil)
+	if err != nil {
+		return nil, err
+	}
+	if amis, ok := p.ec2Cache.Get(fmt.Sprint(hash)); ok {
+		return amis.([]*ec2.Image), nil
+	}
+	// This API is not paginated, so a single call suffices.
+	output, err := p.ec2api.DescribeImagesWithContext(ctx, &ec2.DescribeImagesInput{Filters: filters})
+	if err != nil {
+		return nil, fmt.Errorf("describing images %+v, %w", filters, err)
+	}
+	p.ec2Cache.SetDefault(fmt.Sprint(hash), output.Images)
+	logging.FromContext(ctx).Debugf("Discovered images: %s", amiIDs(output.Images))
+	return output.Images, nil
+}
+
+func getFilters(amiSelector map[string]string) []*ec2.Filter {
+	filters := []*ec2.Filter{}
+	for key, value := range amiSelector {
+		if key == "aws-ids" {
+			filterValues := functional.SplitCommaSeparatedString(value)
+			filters = append(filters, &ec2.Filter{
+				Name:   aws.String("image-id"),
+				Values: aws.StringSlice(filterValues),
+			})
+		} else {
+			filters = append(filters, &ec2.Filter{
+				Name:   aws.String(fmt.Sprintf("tag:%s", key)),
+				Values: []*string{aws.String(value)},
+			})
+		}
+	}
+	return filters
+}
+
+func (p *AMIProvider) getRequirementsFromImage(ec2Image *ec2.Image) scheduling.Requirements {
+	requirements := scheduling.NewRequirements()
+	for _, tag := range ec2Image.Tags {
+		if v1alpha5.WellKnownLabels.Has(*tag.Key) {
+			requirements.Add(scheduling.Requirements{*tag.Key: sets.NewSet(*tag.Value)})
+		}
+	}
+	// Always add the architecture of an image as a requirement, irrespective of what's specified in EC2 tags.
+	architecture := *ec2Image.Architecture
+	if value, ok := awsv1alpha1.AWSToKubeArchitectures[architecture]; ok {
+		architecture = value
+	}
+	requirements.Add(scheduling.Requirements{v1.LabelArchStable: sets.NewSet(architecture)})
+	return requirements
+}
+
+func amiIDs(amis []*ec2.Image) []string {
```

Comment:
optional: depending on your proclivity for functional programming, and could be done inline above.

`lo.Map(amis, func(ami *ec2.Image) string { return ami.ImageId })`

Reply:
Nice!